src/Object/Email/ImageUrl.php

```diff
@@ -0,0 +1,186 @@
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @author Nosto Solutions Ltd <contact@nosto.com>
+ * @copyright 2017 Nosto Solutions Ltd
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD 3-Clause
+ *
+ */
+
+namespace Nosto\Object\Email;
+
+use Nosto\Nosto;
+use Nosto\NostoException;
+
+/**
+ * ImageUrl object formatting email widget image url
+ */
+class ImageUrl
+{
+    const NOSTO_ACCOUNT_PLACEHOLDER = '@NOSTO_ACCOUNT@';
+    const EMAIL_PLACEHOLDER = '@EMAIL@';
+    const RECOMMENDATION_PLACEHOLDER = '@RECOMMENDATION@';
+    const URL_TEMPLATE = 'url_template';
+    const CUSTOMER_EMAIL = 'customer_email';
+    const RECOMMENDATION = 'recommendation';
+    const DEFAULT_URL_TEMPLATE_SUFFIX = '/image/v1/@NOSTO_ACCOUNT@/@RECOMMENDATION@/1?uid=@EMAIL@&version=2.0.8';
+    const DEFAULT_RECOMMENDATION = 'BestSeller';
+
+    /**
+     * @var string|null $urlTemplate If it == null, a EmailWidgetHelper::DEFAULT_URL_TEMPLATE is used.
+     */
+    private $urlTemplate;
+    /**
+     * @var string $nostoAccount nosto account id
+     */
+    private $nostoAccount;
+    /**
+     * @var string $customerEmail customer email address
+     */
+    private $customerEmail;
+    /**
+     * @var string|null $recommendationType recommendation type.
+     * If it == null, ImageUrl::DEFAULT_RECOMMENDATION is used
+     */
+    private $recommendationType;
+
+    /**
+     * ImageUrl constructor.
+     * @param null|string $urlTemplate
+     * @param string $nostoAccount
+     * @param string $customerEmail
+     * @param null|string $recommendationType
+     */
+    public function __construct($urlTemplate, $nostoAccount, $customerEmail, $recommendationType)
```

Comment:
Use `NostoAccountInterface` for `$nostoAccount`

Reply:
If we don't need other information from account, why would we ask for bigger dependency? Smaller dependency makes it easier for the caller.

Reply:
We will probably sign the widget URLs at some point which means are going to need the tokens as well. Also if you pass the account interface the caller doesn't have to care which fields are actually used for URL generation. This works for now but the extendability is poor